### PR TITLE
Convergence diagnostics

### LINKED ISSUE #45: add plotting functions to API
Per MT's suggestion

### LINKED ISSUE #46: remove all "sol.<blah>" from make_figs
Don't pass sol to make_figs, just pass the dict entries in sol that are needed to plot.

### LINKED ISSUE #47: review all remaining inline "# TODO"s
In .py files and docs

### LINKED ISSUE #49: Add diagnostic plots for convergence
Following discussion on Wed afternoon, add two diagnostic plots to the automatically generated figures.
Purpose of these figures is to inform the user on the convergence status of the fit.

### LINKED ISSUE #50: Documentation for diagnostic figures of convergence
After implementing the figures in #49, add a short description/comment on how to interpret these figures to the documentation.

### PR DESCRIPTION
**Pull request by jeffjennings**
Add plotting for convergence diagnostics, other various minor changes

Fix #45 Fix #46 Fix #47 Fix #49 Fix #50

@jeffjennings: the "Fix #N" statements I added yesterday in this description were not reminders to be removed.

A "Fix #N" in a PR description or in a commit message will close the #N issue when the PR is merged. It is a bad practice to close the issues manually, and it is actually better not to unless they become irrelevant, and won't imply any change to the repo. 
Using the "Fix #N" statements ensures that the issues are closed really when they are closed (i.e. when the PR is merged). Until this PR is merged the issues are not closed. If we were not to merge this PR now, you would have closed issues that are not actually closed, with the obvious confusion. 
Secondly, by using "Fix #N" statements Github automatically adds a reference with a link to the PR or the commit where the issue was closed, in the issue page, which helps a lot debugging later on. While now you added manually "see file..." without the PR reference number, which makes it very hard to track the commit down

### CODE REVIEW

**Comment by rbooth200** [comment]
This PR is ready to go from my point of view.

**Comment by rbooth200** [comment]
Marco tells me @jeffjennings is still working on this, I'll wait to close it.

**Comment by rbooth200** [comment]
I've fixed some issues that pyling was giving, but the diagnostic plots aren't working yet, apparently.

**Comment by rbooth200** [comment]
Quick Q: why did you remove the pylint directives?

**Comment by jeffjennings** [comment]
My mistake, I thought it had automatically added those and that they could be removed

**Comment by jeffjennings** [comment]
Everything works for me on this branch now, including the diagnostic plot.

**Comment by rbooth200** [comment]
I'd also suggest replacing start_iter and stop_iter with a single num_iter parameter and just print the so many iterations. Isn't that what will be wanted anyway? I think there is too much chance of error with the current parameters. 

Obviously if you ask for 1000 and it converged in 20 then we just show all of them.

**Comment by rbooth200** [comment]
The pylint directives are to stop it reporting false positives

**Comment by jeffjennings** [comment]
yes good point, I just pushed with update

**Comment by rbooth200** [comment]
That's not quite what I had in mind. Ill

**Comment by jeffjennings** [comment]
I just updated again

**Comment by rbooth200** [comment]
Sure, but i was thinking of changing how those parameters work

**Comment by jeffjennings** [comment]
Open to another way, but I think the current `iter_plot_range` is good; an error isn't raised if it's specified inconsistently with `max_iter`, it's just handled by frank and printed to terminal.

**Comment by rbooth200** [comment]
Ok, I'll just see if I can make it less error prone.

**Comment by jeffjennings** [comment]
On TODOs, did you add the TODO in `fit.py`, l.35,

`frank_path = os.path.dirname(frank.__file__)  # TODO`

and if so is the TODO still needed? 

I also don't understand `fit.py`, l.96,
```
        uv_path = model['input_output']['uvtable_filename']
        if uv_path: 
```
becaue `uv_path` will always be true. should `if uv_path` be unindented?

**Comment by rbooth200** [comment]
I'll look at the Todo when I get in.

The uv_path check is to protect against empty uv table strings. Maybe dirname works properly in that case, I don't know. 

I understand that an empty UV path is an error, but you don't want the code to break at this location if the path is empty, so I just chose a sensible option.

**Comment by rbooth200** [comment]
You can remove the if statement if you think that it is unnecessary, but it produces the correct behaviour as is

**Comment by jeffjennings** [comment]
`frank_path = os.path.dirname(frank.__file__) # TODO` is the last TODO currently in the code, so feel free to close #47 when resolved.

**Comment by jeffjennings** [comment]
> Ok, I'll just see if I can make it less error prone.

I think I see what you mean, I've rewritten and will push in a minute after testing.

**Comment by jeffjennings** [comment]
just fyi your autopep8 isn't using 80 characters as the maximum line length, so when it increases the indents in some blocks, it can make lines fairly longer when they have spaces between variables where a new line could start.

**Comment by rbooth200** [comment]
Can you look at my last commit. I'm now using `None` to specify a default behaviour of the full range. Otherwise its up to the user to specify a valid range

**Comment by jeffjennings** [comment]
Looking at it now; please don't run autopep8 on the .json files; it screws up the readability.

**Comment by jeffjennings** [comment]
It also makes the indenting on the brackets incorrect.

**Comment by rbooth200** [comment]
> Looking at it now; please don't run autopep8 on the .json files; it screws up the readability.

It not autopep8, but anyway, I've turned off the auto formatting.

**Comment by jeffjennings** [comment]
This makes less sense than what was there before. 

- Don't call the diag plot just to plot 0 iterations. 

- If the user can set `max_iter`in the .json, they can specify `iter_plot_range` ... just set the upper bound to large (say 2000) by default, and anyway I print the range being plotted when `make_diag_fig` is called.

**Comment by rbooth200** [comment]
>This makes less sense than what was there before.
>
> -   Don't call the diag plot just to plot 0 iterations.
>
>   -  If the user can set max_iterin the .json, they can specify iter_plot_range ... just set the upper bound to large (say 2000) by default, and anyway I print the range being plotted when make_diag_fig is called.

Actually, I think its sensible to call the plot with no iterations if that is consistent with what was requrested. The figure will be set up with correct axis-labels etc, so the user will at least see what should have been there.

**Comment by rbooth200** [comment]
I also think its better for the code to complete, doing something sensible, but without the diagnostic plot if a bad iteration range was chosen. This way the fit will be stored, along with the diagnostics so the user can get at them without running the code again, if they want.

**Comment by rbooth200** [comment]
> just fyi your autopep8 isn't using 80 characters as the maximum line length, so when it increases the indents in some blocks, it can make lines fairly longer when they have spaces between variables where a new line could start.

I'll fix that...

**Comment by jeffjennings** [comment]
> > This makes less sense than what was there before.
> > 
> > * Don't call the diag plot just to plot 0 iterations.
> > * If the user can set max_iterin the .json, they can specify iter_plot_range ... just set the upper bound to large (say 2000) by default, and anyway I print the range being plotted when make_diag_fig is called.
> 
> Actually, I think its sensible to call the plot with no iterations if that is consistent with what was requrested. The figure will be set up with correct axis-labels etc, so the user will at least see what should have been there.

The problem is that alternatively it could just look like the code is broken and doesn't populate the plot that it's supposed to, and the user is confused why (and actually, if the plotting is broken in some other way, that's then degenerate with this). I don't think it makes sense to plot an empty figure; just explicitly state in the logging message printed to the terminal that the plot won't be made. You don't have to raise an error to stop the code, just print the log message and skip the diag plot.

**Comment by rbooth200** [comment]
Isn't that what the logging message is for?

**Comment by rbooth200** [comment]
You don't want to skip the full diagnostic plot because the ranges are different for different sub panels

**Comment by jeffjennings** [comment]
> You don't want to skip the full diagnostic plot because the ranges are different for different sub panels

That's incorrect, all plots in the diagnostic fig are produced using `iter_plot_range`

If you make that parameter `null` by default, it's also not immediately clear it needs to be a list. I'm setting the default as 
```
"iter_plot_range"  : [0,5000]
```
so that for 99% of cases it won't matter.

**Comment by rbooth200** [comment]
Actually, you only use the last 100 iterations for 2 of the panels

**Comment by jeffjennings** [comment]
Hey @mtazzari, @rbooth200  and I have both been contributing to this branch that closes the outstanding issues in the Code freeze milestone. Can you review and merge into master? Thanks!

**Comment by mtazzari** [comment]
Ok, starting to review

**Comment by rbooth200** [comment]
I've just merged in some changes needed to get tests of the full pipeline working

File: frank/geometry.py [hunk 1]
```diff
@@ -22,12 +22,10 @@
    NOTE: The sign convention used here is xx.
 """
 
-import numpy as np
```
Comment (mtazzari) [inline frank/geometry.py]: In general, I see that it was adopted a convention to import `frank` modules first and external modules (e.g., `numpy`, `scipy`, ...) later.
Actually, the [correct PEP8 practice](https://www.python.org/dev/peps/pep-0008/#imports) is the other way around: 
```
    1. Standard library imports.
    2. Related third party imports.
    3. Local application/library specific imports.
```
One has to ensure frank is compliant with the distribution, not the other way around.
Hopefully `np`, or `least_squares` are not objects in `frank`, but if it was the case, `import numpy as np` as it is right now would shadow our `np` object.
So, nothing to change for now as long as it works as intended.
In the future, we might just want to adopt the right convention in the whole codebase.

File: frank/radial_fitters.py [hunk 2]
```diff
@@ -20,16 +20,14 @@
    of deprojected visibities.
 """
 
+from frank.hankel import DiscreteHankelTransform
```
Comment (mtazzari) [inline frank/radial_fitters.py]: here another case of importing frank before standard packages. risky

File: docs/install.rst [hunk 3]
```diff
@@ -24,11 +24,13 @@ Clone the source repository from `GitHub <https://github.com/discsim/frank>`_,
 
     pip install git+https://github.com/discsim/frank.git
 
+(this is the same as `git clone https://github.com/discsim/frank.git; cd frank`).
```
Comment (mtazzari) [inline docs/install.rst]: This is not the same, actually the installation command is missing.
```suggestion
(this is the same as `git clone https://github.com/discsim/frank.git && cd frank && pip install .`).
```

File: frank/geometry.py (hunk 1 above)
Comment (jeffjennings) [reply frank/geometry.py]: Good to note, thanks!

File: frank/radial_fitters.py (hunk 2 above)
Comment (jeffjennings) [reply frank/radial_fitters.py]: cool, we'll do this in another PR!

**Pull request merged**